http4k-format/moshi/src/test/kotlin/org/http4k/format/MoshiTest.kt

```diff
@@ -42,9 +44,128 @@ class MoshiAutoTest : AutoMarshallingJsonContract(Moshi) {
+        ))
+
+        val json = with(Moshi) {
+            element.asCompactJsonString()
+        }
+
+        assertThat(json, equalTo("""{"string":"hello","child":{"string":"world","numbers":[1,1.2],"bool":true},"numbers":[],"bool":false}"""))
+    }
+
+    @Test
+    fun `convert arbObject to MoshiElement`() {
+        assertThat(
+            Moshi.asJsonObject(obj),
+            equalTo(
+                MoshiObject(mapOf(
+                    "string" to MoshiString("hello"),
+                    "child" to MoshiObject(mapOf(
+                        "string" to MoshiString("world"),
+                        "numbers" to MoshiArray(listOf(
+                            MoshiInteger(BigInteger.valueOf(1))
+                        )),
+                        "bool" to MoshiBoolean(true)
+                    )),
+                    "numbers" to MoshiArray(emptyList()),
+                    "bool" to MoshiBoolean(false)
+                ))
+            )
+        )
+    }
+
+    @Test
+    fun `convert MoshiElement to arbObject`() {
+        val element = MoshiObject(mapOf(
+            "string" to MoshiString("hello"),
+            "child" to MoshiObject(mapOf(
+                "string" to MoshiString("world"),
+                "numbers" to MoshiArray(listOf(
+                    MoshiInteger(BigInteger.valueOf(1))
+                )),
+                "bool" to MoshiBoolean(true)
+            )),
+            "numbers" to MoshiArray(emptyList()),
+            "bool" to MoshiBoolean(false)
+        ))
+
+        assertThat(
+            Moshi.asA(element, ArbObject::class),
+            equalTo(obj)
+        )
+    }
+
     override fun customMarshaller() = object : ConfigurableMoshi(Builder().asConfigurable().customise()) {}
     override fun customMarshallerProhibitStrings() = object : ConfigurableMoshi(
         Builder().asConfigurable().prohibitStrings()
             .customise()
     ) {}
 }
+
+class MoshiJsonTest: JsonContract<MoshiNode>(Moshi) {
+    override val prettyString = """{
+    "hello": "world"
+}"""
+
+    @Test
+    override fun `serializes object to json`() {
+        j {
+            val input = obj(
+                "string" to string("value"),
+                "double" to number(1.5),
+                "long" to number(10L),
+                "boolean" to boolean(true),
+                "bigDec" to number(BigDecimal(1.2)),
+                "bigInt" to number(BigInteger("12344")),
+                "null" to nullNode(),
+                "int" to number(2),
+                "empty" to obj(),
+                "array" to array(listOf(
+                    string(""),
+                    number(123)
+                )),
+                "singletonArray" to array(obj("number" to number(123)))
+            )
+            val expected = """{"string":"value","double":1.5,"long":10,"boolean":true,"bigDec":1.2,"bigInt":12344,"int":2,"empty":{},"array":["",123],"singletonArray":[{"number":123}]}"""
```

Comment:
Difference here is that the `bigDec` doesn't suffer from the floating-point imprecision of the original, and `null` is missing, because moshi doesn't serialize nulls, by default.